storage/pin/pin.go

```diff
@@ -0,0 +1,408 @@
+	}
+
+	// Store the pinned files in state DB
+	err = p.state.Put("pin-files", p.pinnedFiles)
+	if err != nil {
+		log.Error("Error pinning file.", "rootHash", rootHash, "err", err)
+		return nil
+	}
+
+	log.Debug("File pinned", "Address", rootHash)
+	return nil
+}
+
+// UnPinFiles is used to unpin an already pinned file. It takes the root
+// hash of the file and walks down the merkle tree unpinning all the chunks
+// that are encountered on the way. The pre-requisite is that the file should
+// have been already pinned using the PinFiles function. This function can
+// be called only from an external command.
+func (p *API) UnpinFiles(rootHash string, credentials string) error {
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash", "rootHash", rootHash, "err", err)
+		return err
+	}
+
+	p.pinLock.RLock()
+	fileInfo, ok := p.pinnedFiles[rootHash]
+	p.pinLock.RUnlock()
+	if !ok {
+		log.Error("Root hash is not pinned", "rootHash", rootHash)
+		return err
+	}
+
+	// Walk the root hash and unpin all the chunks
+	walkerFunction := func(ref storage.Reference) error {
+		chunkAddr := p.removeDecryptionKeyFromChunkHash(ref)
+		err := p.db.Set(context.Background(), chunk.ModeSetUnpin, chunkAddr)
+		if err != nil {
+			log.Error("Could not unpin chunk", "Address", hex.EncodeToString(chunkAddr))
+			return err
+		} else {
+			log.Trace("Unpinning chunk", "Address", hex.EncodeToString(chunkAddr))
+		}
+		return nil
+	}
+	p.walkChunksFromRootHash(rootHash, fileInfo.IsRaw, credentials, walkerFunction)
+
+	// Update the state DB
+	p.pinLock.Lock()
+	defer p.pinLock.Unlock()
+	pinCounter, err := p.getPinCounterOfChunk(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+	if err != nil {
+		delete(p.pinnedFiles, rootHash)
+	} else {
+		fileInfo.PinCounter = pinCounter
+		p.pinnedFiles[rootHash] = fileInfo
+	}
+
+	// Update the unpinned files in state DB
+	err = p.state.Put("pin-files", p.pinnedFiles)
+	if err != nil {
+		log.Error("Error unpinning file.", "rootHash", rootHash, "err", err)
+		return nil
+	}
+
+	log.Debug("File unpinned", "Address", rootHash)
+	return nil
+}
+
+// ListPinFiles functions logs information of all the files that are pinned
+// in the current local node. It displays the root hash of the pinned file
+// or collection. It also display three vital information's
+//     1) Whether the file is a RAW file or not
+//     2) Size of the pinned file or collection
+//     3) the number of times that particular file or collection is pinned.
+func (p *API) ListPinFiles() map[string]FileInfo {
+	p.pinLock.RLock()
+	defer p.pinLock.RUnlock()
+	for hash, fileInfo := range p.pinnedFiles {
+		log.Info("Pinned file", "Address", hash, "IsRAW", fileInfo.IsRaw,
+			"FileSize", fileInfo.FileSize, "PinCounter", fileInfo.FileSize)
+	}
+	return p.pinnedFiles
+}
+
+func (p *API) walkChunksFromRootHash(rootHash string, isRaw bool, credentials string, executeFunc func(storage.Reference) error) {
```

Comment:
would be nice to handle `rootHash` as `chunk.Address`

Reply:
This is called from multiple places.. so instead of doing it there.. i do it once here...